### PR TITLE
API/DEPR: 'periods' argument instead of 'n' for DatetimeIndex.shift() 

### LINKED ISSUE #22458: Inconsistent keyword argument name for the 'periods' in pandas.Index.shift for a DatetimeIndex
The actual implementation of ``pandas.Index.shift`` for a datetime-like index takes a differently named parameter for the number of shifts than the base method.

It should be named ``periods``, which would also be consistent with the ``pandas.DataFrame.shift`` or ``pandas.Series.shift`` methods. But for a DatetimeIndex it's ``n``.

See:
https://github.com/pandas-dev/pandas/blob/v0.23.4/pandas/core/indexes/base.py#L2586
vs.
https://github.com/pandas-dev/pandas/blob/v0.23.4/pandas/core/indexes/datetimelike.py#L1021

With a DatetimeIndex, following the documentation leads to a
```
TypeError: shift() got an unexpected keyword argument 'periods'
```

If this is somehow intended, then it should be stated so in the docstring for ``pandas.Index.shift``.

pandas version is 0.23.4

### PR DESCRIPTION
**Pull request by arminv**
- [x] closes #22458 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

In order to be consistent with `Index.shift` & `Series.shift`, `n` argument was deprecated in favor of `periods`.

```
In [2]: idx = pd.DatetimeIndex(start='2014-08-01 10:00', freq='H',
   ...:                         periods=3, tz='Asia/Calcutta')
   ...:                         

In [3]: idx
Out[3]: 
DatetimeIndex(['2014-08-01 10:00:00+05:30', '2014-08-01 11:00:00+05:30',
               '2014-08-01 12:00:00+05:30'],
              dtype='datetime64[ns, Asia/Calcutta]', freq='H')

In [4]: idx.shift(1, freq=None)
Out[4]: 
DatetimeIndex(['2014-08-01 11:00:00+05:30', '2014-08-01 12:00:00+05:30',
               '2014-08-01 13:00:00+05:30'],
              dtype='datetime64[ns, Asia/Calcutta]', freq='H')

In [5]: idx.shift(periods=1, freq=None)
Out[5]: 
DatetimeIndex(['2014-08-01 11:00:00+05:30', '2014-08-01 12:00:00+05:30',
               '2014-08-01 13:00:00+05:30'],
              dtype='datetime64[ns, Asia/Calcutta]', freq='H')

In [6]: idx.shift(n=1, freq=None)
//anaconda/envs/pandas_dev/bin/ipython:1: FutureWarning: the 'n' keyword is deprecated, use 'periods' instead
  #!//anaconda/envs/pandas_dev/bin/python
Out[6]: 
DatetimeIndex(['2014-08-01 11:00:00+05:30', '2014-08-01 12:00:00+05:30',
               '2014-08-01 13:00:00+05:30'],
              dtype='datetime64[ns, Asia/Calcutta]', freq='H')
```

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 1]
```diff
@@ -540,6 +540,12 @@ def test_shift(self):
             shifted = rng.shift(1, freq=CDay())
             assert shifted[0] == rng[0] + CDay()
 
+    def test_shift_periods(self):
```
Comment (arminv) [inline pandas/tests/indexes/datetimes/test_ops.py]: Not sure if this is the right way to test this

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -521,40 +521,52 @@ def _addsub_offset_array(self, other, op):
             kwargs['freq'] = 'infer'
         return type(self)(res_values, **kwargs)
 
-    def shift(self, n, freq=None):
+    def shift(self, periods, freq=None):
```
[the remark this answers is not in the document]
Comment (arminv) [reply pandas/core/arrays/datetimelike.py]: @TomAugspurger Thanks. I started parsing manually but couldn't make some tests pass. Then I realized there is a decorator in `pandas/pandas/util/_decorators.py` that I could use for deprecating `n`. Lmk if there is anything else to be done.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -528,6 +528,7 @@ Datetimelike API Changes
 - :class:`PeriodIndex` subtraction of another ``PeriodIndex`` will now return an object-dtype :class:`Index` of :class:`DateOffset` objects instead of raising a ``TypeError`` (:issue:`20049`)
 - :func:`cut` and :func:`qcut` now returns a :class:`DatetimeIndex` or :class:`TimedeltaIndex` bins when the input is datetime or timedelta dtype respectively and ``retbins=True`` (:issue:`19891`)
 - :meth:`DatetimeIndex.to_period` and :meth:`Timestamp.to_period` will issue a warning when timezone information will be lost (:issue:`21333`)
+- :func:`shift` now accepts ``periods`` argument instead of ``n`` for consistency with `Index.shift`. Using ``n`` throws a deprecation warning (:issue:`22458`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: I don't think ``:func:`shift` `` will render, you mean ``:func:`Series.shift` ``?, also annotate ``Index.shift`` as well.

Moves this to the deprecation section

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 4]
```diff
@@ -540,6 +540,16 @@ def test_shift(self):
...
+        # GH #22458 :
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: can you mention that n is deprecated here in a comment

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -521,40 +522,51 @@ def _addsub_offset_array(self, other, op):
...
+        by a specified time increment a given number of times.
 
         Parameters
         ----------
-        n : int
-            Periods to shift by
-        freq : DateOffset or timedelta-like, optional
+        periods : int
+            Number of periods (or increments) to shift by,
+            can be positive or negative.
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: versionchanged -> 0.24.0

File: pandas/core/arrays/datetimelike.py [hunk 6]
```diff
@@ -521,40 +522,51 @@ def _addsub_offset_array(self, other, op):
...
+            Offset aliases are valid strings, e.g., 'D', 'W', 'M' etc.
 
         Returns
         -------
-        shifted : same type as self
+        pandas.DatetimeIndex
+            Shifted index.
+
+        See Also
+        --------
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you edit the Series.shift to add a See Also back to here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py [hunk 7]
```diff
@@ -8298,6 +8298,11 @@ def mask(self, cond, other=np.nan, inplace=False, axis=None, level=None,
             See Notes.
         axis : %(axes_single_arg)s
 
+        See Also
+        --------
+        Index.shift : Shift values of Index.
+        DatetimeIndex.shift : Shift values of DatetimeIndex.
+
```
Comment (arminv) [inline pandas/core/generic.py]: @jreback I also added `Index.shift` here.

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 8]
```diff
@@ -540,6 +540,16 @@ def test_shift(self):
...
             assert shifted[0] == rng[0] + CDay()
 
+    def test_shift_periods(self):
+        # GH #22458 : argument 'n' was deprecated in favor of 'periods'
+        idx = pd.DatetimeIndex(start=START, end=END,
+                               periods=3)
+        tm.assert_index_equal(idx.shift(periods=0), idx)
+        tm.assert_index_equal(idx.shift(0), idx)
+        with tm.assert_produces_warning(FutureWarning,
+                                        check_stacklevel=False):
```
Comment (TomAugspurger) [inline pandas/tests/indexes/datetimes/test_ops.py]: Could you change this to true? If the test fails, you may need to increase the stacklevel 1.

Comment (arminv) [reply pandas/tests/indexes/datetimes/test_ops.py]: @TomAugspurger done, test passed too.

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**